### PR TITLE
Optimize pserver thread pool

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: paddle/fluid/operators/distributed/request_handler_impl.cc [hunk 1]
```diff
@@ -42,13 +43,16 @@ bool RequestSendHandler::Handle(const std::string& varname,
...
   if (!sync_mode_) {
     rpc_server_->Profiler().OneStep();
-    try {
-      executor_->RunPreparedContext((*grad_to_prepared_ctx_)[varname].get(),
-                                    scope);
-    } catch (std::exception& e) {
-      LOG(ERROR) << "async: run sub program error " << e.what();
-      return false;
-    }
+    std::future<void> ft = framework::Async([this, &scope, varname]() {
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/request_handler_impl.cc]: Maybe add one comment that why need to run in thread here.

**Review by typhoonzero — DISMISSED** [review]

**Comment by jacquesqiao** [comment]
# Before update
#### Prefetchs are running in prefetch rpc thread(5) while optimization run in thread pool.

image

# After put prefetch into thread pool
#### prefetch and optimization now all run in thread pool, because we only have 5 prefetch RPC thread, prefetch can only use 5 thread at the same time.
image

# After set the prefetch rpc thread num to thread_pool_size * 2
#### prefetch can use all thread in the thread_pool now.
image

File: paddle/fluid/operators/distributed/request_handler_impl.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/request_handler_impl.cc]: done

**Comment by panyx0718** [comment]
What's the speed improvement

**Comment by panyx0718** [comment]
Can you only change prefetch? I remember I tune this number before. more is not always better

File: paddle/fluid/operators/distributed/request_handler_impl.cc [hunk 2]
```diff
@@ -129,13 +137,15 @@ bool RequestCheckpointHandler::Handle(const std::string& varname,
...
-  executor_->RunPreparedContext(checkpoint_prepared_ctx_.get(), scope);
+  std::future<void> ft = framework::Async([this, &scope, varname]() {
+    auto* lt_var = scope->FindVar(LOOKUP_TABLE_PATH)->GetMutable<std::string>();
+    lt_var->clear();
+    lt_var->append(out_var_name);
+    VLOG(4) << "RequestCheckpointHandler update var kLookupTablePath to: "
+            << out_var_name;
+    executor_->RunPreparedContext(checkpoint_prepared_ctx_.get(), scope);
+  });
+  ft.wait();
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/request_handler_impl.cc]: why is this faster?

**Review by panyx0718 — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 3]
```diff
@@ -275,14 +275,17 @@ void ListenAndServOp::RunImpl(const framework::Scope &scope,
   request_checkpoint_handler_.reset(new distributed::RequestCheckpointHandler(
       sync_mode, checkpoint_block_id));
 
+  int thread_pool_size =
+      static_cast<int>(framework::ThreadPool::GetInstance()->Threads());
   rpc_service_->RegisterRPC(distributed::kRequestSend,
-                            request_send_handler_.get());
+                            request_send_handler_.get(), thread_pool_size * 2);
```
Comment (typhoonzero) [inline paddle/fluid/operators/listen_and_serv_op.cc]: Maybe create only one list of threads inside `rpc_service_` is enough? in `grpc_server.cc:285`

File: paddle/fluid/operators/distributed/request_handler_impl.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/request_handler_impl.cc]: Currently, each RPC can only use 5 thread, but the pserver can have more CPUs, such as 12, we can use more CPU if put the computation into the thread pool, and can control the size of the thread pool to optimize the computation.

**Review by typhoonzero — APPROVED** [review]
LGTM++

File: paddle/fluid/operators/distributed/request_handler_impl.cc [hunk 4]
```diff
@@ -113,11 +118,15 @@ bool RequestPrefetchHandler::Handle(const std::string& varname,
...
-
+  // All execution should use thread in the thread pool to do.
+  std::future<void> ft =
+      framework::Async([this, &scope, &varname, &out_var_name, &outvar]() {
+        auto var_desc = program_->Block(0).FindVar(out_var_name);
+        InitializeVariable(*outvar, var_desc->GetType());
+        executor_->RunPreparedContext(
+            (*prefetch_var_name_to_prepared_ctx_)[varname].get(), scope);
+      });
+  ft.wait();
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/request_handler_impl.cc:129]: I still suspect this pattern really works. What if `RequestPrefetchHandler::Handle` is put in a threadpool? Then it's thread in thread?

**Review by panyx0718 — APPROVED** [review]

File: paddle/fluid/operators/distributed/request_handler_impl.cc (hunk 4 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/request_handler_impl.cc:129]: If `RequestPrefetchHandler::Handle` is in thread_pool, then this change is not needed.

**Pull request closed without merging**